README.md
ProSGPV
========
Penalized Regression with Second-Generation P-Values

- [1. Introduction](#1-introduction)
- [2. Installation](#2-installation)
- [3. Example](#3-example)
  * [3.1 How ProSGPV works](#31-how-prosgpv-works)
  * [3.2 Real-world-data](#32-real-world-data)
    + [3.2.1 Two-stage-algorithm](#321-two-stage-algorithm)
    + [3.2.2 One-stage-algorithm](#322-one-stage-algorithm)
  * [3.3 More-examples](#33-more-examples)
- [4. References](#4-references)

<!-- toc -->

# 1. Introduction

We know that p-values can't be used for variable selection. However, you can do so with second-generation p-values. Here is how.

# 2. Installation

To install it on CRAN, you can do

``` r
install.packages("ProSGPV")
```

For a development version of ProSGPV, you can install it by running the following command.  

``` r
library(devtools)
devtools::install_github("zuoyi93/ProSGPV")
```

# 3. Example

## 3.1 How ProSGPV works

Below is an illustration of how ProSGPV successfully selects the true support, while lasso and fully relaxed lasso fails. Five variables are simulated and only V3 is associated with the response. Plot (1) presents the lasso solution path. The vertical dotted line is the lambda selected by generalized information criterion. (2) shows the fully relaxed lasso path. (3) shows the fully relaxed lasso paths with their 95% confidence intervals (in lighter color). (4) illustrates the ProSGPV algorithm selection path. The shaded area is the null region; the colored lines are each 95% confidence bound that is closer to the null region. Lasso and fully relaxed lasso would select both V5 and V3, while ProSGPV successfully screens out V2.  

![](vignettes/assets/works.png)

## 3.2 Real-world data

Here, we use the Tehran housing data as an illustrative real-world data example of how ProSGPV works with linear regression. 

The Tehran housing data contain 26 explanatory variables and one outcome. Details about data collection can be found in this [paper](https://ascelibrary.org/doi/abs/10.1061/%28ASCE%29CO.1943-7862.0001047), and variable description can be found [here](man/t.housing.Rd). 

### 3.2.1 Two-stage algorithm 

We can load the Tehran Housing data `t.housing` stored in the package.

``` r
# load the package
library(ProSGPV)

# prepare the data
x = t.housing[,-ncol(t.housing)]
y = t.housing$V9

# run ProSGPV
out.sgpv.2 <- pro.sgpv(x = x, y = y)
```
The two-stage algorithm selects the following variables.

``` r
out.sgpv.2
```

    #> Selected variables are V8 V12 V13 V15 V17 V26

We can view the summary of the final model.

```r
summary(out.sgpv.2)
```
	#> 
	#> Call:
	#> lm(formula = Response ~ ., data = data.d)
	#> 
	#> Residuals:
	#>      Min       1Q   Median       3Q      Max 
	#> -1276.35   -75.59    -9.58    59.46  1426.22 
	#> 
	#> Coefficients:
	#>               Estimate Std. Error t value Pr(>|t|)    
	#> (Intercept)  1.708e+02  3.471e+01   4.920 1.31e-06 ***
	#> V8           1.211e+00  1.326e-02  91.277  < 2e-16 ***
	#> V12         -2.737e+01  2.470e+00 -11.079  < 2e-16 ***
	#> V13          2.185e+01  2.105e+00  10.381  < 2e-16 ***
	#> V15          2.041e-03  1.484e-04  13.756  < 2e-16 ***
	#> V17         -3.459e+00  8.795e-01  -3.934  0.00010 ***
	#> V26         -4.683e+00  1.780e+00  -2.630  0.00889 ** 
	#> ---
	#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
	#> 
	#> Residual standard error: 194.8 on 365 degrees of freedom
	#> Multiple R-squared:  0.9743, Adjusted R-squared:  0.9739 
	#> F-statistic:  2310 on 6 and 365 DF,  p-value: < 2.2e-16


Coefficient estimates can be extracted by use of `coef`.

```r
coef(out.sgpv.2s)
```

	#>  [1]   0.000000000   0.000000000   0.000000000   0.000000000   0.000000000
	#>  [6]   0.000000000   1.210755031   0.000000000 -27.367601037  21.853920174
	#> [11]   0.000000000   0.002040784   0.000000000  -3.459496972   0.000000000
	#> [16]   0.000000000   0.000000000   0.000000000   0.000000000   0.000000000
	#> [21]   0.000000000   0.000000000  -4.683172725   0.000000000   0.000000000
	#> [26]   0.000000000

In-sample prediction can be made using S3 method `predict` and an external sample can be provided to make out-of-sample prediction with an argument of `newx` in the predict function.

```r
head(predict(out.sgpv.2s))
```

	#>         1         2         3         4         5         6 
	#> 1565.7505 3573.7793  741.7576  212.1297 5966.1682 5724.0172


S3 method `plot` can be used to visualize the variable selection process.

``` r
plot(out.sgpv.2, lambda.max = 0.01)
```
First, we plot the full solution path with point estimates and 95% confidence intervals. Note that the null region is in grey. The selected variables are colored blue on the y-axis. `lambda.max` controls the limit of the X-axis.  

![](vignettes/assets/linear.fig.1.png)

Alternatively, we can plot the confidence bound that is closer to the null.

``` r
plot(out.sgpv.2,lpv=1,lambda.max=0.01)
```

![](vignettes/assets/linear.fig.2.png)


### 3.2.2 One-stage algorithm

A fast one-stage ProSGPV algorithm is also available when <img src="https://latex.codecogs.com/png.latex?\color{blue}{n>p}" />.

``` r
# run one-stage algorithm
out.sgpv.1 <- pro.sgpv(x = x, y = y, stage = 1)
```

The selected variables are  

``` r
out.sgpv.1
```

    #> Selected variables are V8 V12 V13 V15 V17 V25 V26

`coef`, `summary`, and `predict` methods are also applicable here.

Moreover, `plot` function can be used to visualize the variable selection process for the one-stage algorithm. Point estimates and 95% confidence intervals are shown for each variable, and the null bounds are shown in green vertical bars. Selected variables are colored in blue.

``` r
plot(out.sgpv.1)
```

![](vignettes/assets/linear.fig.3.png)

However, it is recommended to use the two-stage algorithm rather than the one-stage algorithm for better support recovery and parameter estimation. More importantly, only the two-stage algorithm is available for high dimensional data where <img src="https://latex.codecogs.com/png.latex?\color{blue}{p>n}" />.


## 3.3 More examples

For [high-dimensional continuous data examples](https://zuoyi93.github.io/PDF/Vignette%20ProSGPV%20Linear.html) and [GLM and Cox examples](https://zuoyi93.github.io/PDF/Vignette%20ProSGPV%20GLM%20and%20Cox.html), please refer to the [vignette](vignettes) folder.  


# 4. References

The paper that proposed ProSGPV algorithm in linear regression:  

	Yi Zuo, Thomas G. Stewart & Jeffrey D. Blume (2021) Variable Selection With Second-Generation P-Values, The American Statistician, DOI: 10.1080/00031305.2021.1946150

The paper that proposed ProSGPV algorithm in GLM and Cox models:

	Coming soon

The papers regarding the second-generation p-values:  

	Blume JD, Greevy RA, Welty VF, Smith JR, Dupont WD. An introduction to second-generation p-values. The American Statistician. 2019 Mar 29;73(sup1):157-67.

	Blume JD, D’Agostino McGowan L, Dupont WD, Greevy Jr RA. Second-generation p-values: Improved rigor, reproducibility, & transparency in statistical analyses. PLoS One. 2018 Mar 22;13(3):e0188299.

The paper on finding tuning parameters with generalized information criterion:

	Fan Y, Tang CY. Tuning parameter selection in high dimensional penalized likelihood. Journal of the Royal Statistical Society: SERIES B: Statistical Methodology. 2013 Jun 1:531-52.



NEWS.md
# ProSGPV

Penalized Regression with Second-Generation P-Values


## ProSGPV 1.0.0

8/6/2021: ProSGPV version 1.0.0 is here!

* Added the reference that proposed the ProSGPV algorithm

## ProSGPV 0.3.0

5/4/2021: ProSGPV is faster now!

* Changed the first-stage screening method from cross-validated lasso to one using generalized information criterion

* Removed `which.sgpv` as there is no variability in selection results given a data set

## ProSGPV 0.2.0

3/31/2021: ProSGPV can handle GLM now!

* Added functionalities of variable selection in Logistic, Poisson, and Cox models

* Added the `which.sgpv` function to find the most frequent model when the two-stage algorithm is used

* Added the S3 `plot` method for the one-stage algorithm

* Added vignettes to better illustrate how ProSGPV works

## ProSGPV 0.1.0

1/6/2021: The first version of ProSGPV is released now!

* Added a NEWS.md file to track changes to the package



R/pro.sgpv.R
#' \code{pro.sgpv} function
#'
#' This function outputs the variable selection results
#' from either one-stage algorithm or two-stage algorithm.
#'
#' @importFrom stats complete.cases coef lm predict
#' @importFrom glmnet glmnet
#' @param x Independent variables, can be a \code{matrix} or a \code{data.frame}
#' @param y Dependent variable, can be a \code{vector} or a column from a \code{data.frame}
#' @param stage Algorithm indicator. 1 denotes the one-stage algorithm and
#' 2 denotes the two-stage algorithm. Default is 2. When \code{n} is less than \code{p},
#' only the two-stage algorithm is available.
#' @param family A description of the error distribution and link function to be
#'  used in the model. It can take the value of `\code{gaussian}`, `\code{binomial}`,
#'  `\code{poisson}`, and `\code{cox}`. Default is `\code{gaussian}`
#' @param gvif A logical operator indicating whether a generalized variance inflation factor-adjusted
#' null bound is used. Default is FALSE. See Fox (1992) \doi{10.1080/01621459.1992.10475190}
#' for more details on how to calculate GVIF
#'
#' @return A list of following components:
#' \describe{
#' \item{var.index}{A vector of indices of selected variables}
#' \item{var.label}{A vector of labels of selected variables}
#' \item{lambda}{\code{lambda} selected by generalized information criterion in the two-stage algorithm. \code{NULL} for the one-stage algorithm}
#' \item{x}{Input data \code{x}}
#' \item{y}{Input data \code{y}}
#' \item{family}{\code{family} from the input}
#' \item{stage}{\code{stage} from the input}
#' \item{null.bound}{Null bound in the SGPV screening}
#' \item{pe.can}{Point estimates in the candidate set}
#' \item{lb.can}{Lower bounds of CI in the candidate set}
#' \item{ub.can}{Upper bounds of CI in the candidate set}
#' }
#' @export
#' @seealso
#' * [print.sgpv()] prints the variable selection results
#' * [coef.sgpv()] extracts coefficient estimates
#' * [summary.sgpv()] summarizes the OLS outputs
#' * [predict.sgpv()] predicts the outcome
#' * [plot.sgpv()] plots variable selection results
#' @examples
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # run ProSGPV in linear regression
#' out.sgpv <- pro.sgpv(x = x, y = y)
#'
#' # More examples at https://github.com/zuoyi93/ProSGPV/tree/master/vignettes
pro.sgpv <- function(x, y, stage = c(1, 2),
                     family = c("gaussian", "binomial", "poisson", "cox"),
                     gvif = F) {
  if (!is.numeric(as.matrix(x)) | !is.numeric(y)) stop("The input data have non-numeric values.")

  if (any(complete.cases(x) == F) | any(complete.cases(y) == F)) {
    warning("Only complete records will be used.\n")
    comp.index <- complete.cases(data.frame(x, y))
    if (family != "cox") {
      x <- x[comp.index, ]
      y <- y[comp.index]
    } else {
      x <- x[comp.index, ]
      y <- y[comp.index, ]
    }
  }

  if (missing(stage)) stage <- 2
  if (!stage %in% c(1, 2)) stop("`stage` only takes 1 or 2.")

  if (missing(family)) family <- "gaussian"
  family <- match.arg(family)

  # when p > n, only two-stage is available
  if (stage == 1 & nrow(x) < ncol(x)) stage <- 2

  if (is.null(colnames(x))) colnames(x) <- paste("V", 1:ncol(x), sep = "")

  # standardize inputs in linear regression
  if (family == "gaussian") {
    xs <- scale(x)
    ys <- scale(y)
  } else {
    xs <- as.matrix(x)
    ys <- y
  }

  # get candidate set
  if (stage == 2) {
    temp0 <- get.candidate(xs, ys, family)
    candidate.index <- temp0[[1]]
    lambda <- temp0[[2]]
  } else {
    candidate.index <- 1:ncol(xs)
    lambda <- NULL
  }

  temp <- get.var(candidate.index, xs, ys, family, gvif)
  out.sgpv <- temp[[1]]
  null.bound <- temp[[2]]
  pe.can <- temp[[3]]
  lb.can <- temp[[4]]
  ub.can <- temp[[5]]

  out <- list(
    var.index = out.sgpv,
    var.label = colnames(x)[out.sgpv],
    lambda = lambda,
    x = data.frame(x),
    y = y,
    family = family,
    stage = stage,
    null.bound = null.bound,
    pe.can = pe.can,
    lb.can = lb.can,
    ub.can = ub.can
  )

  class(out) <- "sgpv"
  return(out)
}

#' \code{print.sgpv}: Print variable selection results
#'
#' S3 method \code{print} for an S3 object of class \code{sgpv}
#'
#' @param x An \code{sgpv} object
#' @param ... Other \code{print} arguments
#'
#' @return Variable selection results
#' @export
#' @examples
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # run one-stage algorithm
#' out.sgpv.1 <- pro.sgpv(x = x, y = y, stage = 1)
#'
#' out.sgpv.1
print.sgpv <- function(x, ...) {
  if (length(x$var.index) > 0) {
    cat("Selected variables are", x$var.label, "\n")
  } else {
    cat("None of variables are selected.\n")
  }
}

#' \code{coef.sgpv}: Extract coefficients from the model fit
#'
#' S3 method \code{coef} for an S3 object of class \code{sgpv}
#'
#' @importFrom stats coef lm
#' @importFrom survival coxph Surv
#' @importFrom brglm2 brglmFit
#' @param object An \code{sgpv} object
#' @param ... Other \code{coef} arguments
#'
#' @return Coefficients in the OLS model
#' @export
#' @examples
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # run one-stage algorithm
#' out.sgpv <- pro.sgpv(x = x, y = y)
#'
#' # get coefficients
#' coef(out.sgpv)
coef.sgpv <- function(object, ...) {
  out.coef <- numeric(ncol(object$x))

  if (length(object$var.index) > 0) {
    data.d <- data.frame(yy = object$y, xx = object$x[, object$var.index])

    if (object$family == "gaussian") {
      out.sgpv.coef <- coef(lm(yy ~ ., data = data.d))[-1]
    } else if (object$family == "binomial") {
      out.sgpv.coef <- coef(glm(yy ~ .,
        family = "binomial", data = data.d,
        method = "brglmFit", type = "MPL_Jeffreys"
      ))[-1]
    } else if (object$family == "poisson") {
      out.sgpv.coef <- coef(glm(yy ~ ., family = "poisson", data = data.d))[-1]
    } else {
      cox.m <- coxph(Surv(object$y[, 1], object$y[, 2]) ~
      as.matrix(object$x[, object$var.index]))
      out.sgpv.coef <- coef(cox.m)
    }

    for (i in 1:length(object$var.index)) {
      out.coef[object$var.index[i]] <- out.sgpv.coef[i]
    }
  }
  out.coef
}

#' \code{summary.sgpv}: Summary of the final model
#'
#' S3 method \code{summary} for an S3 object of class \code{sgpv}
#'
#' @importFrom brglm2 brglmFit
#' @param object An \code{sgpv} object
#' @param ... Other arguments
#'
#' @return Summary of a model
#' @export
#' @examples
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # run one-stage algorithm
#' out.sgpv <- pro.sgpv(x = x, y = y)
#'
#' # get regression summary
#' summary(out.sgpv)
summary.sgpv <- function(object, ...) {
  if (length(object$var.index) > 0) {
    if (object$family != "cox") {
      data.d <- data.frame(yy = object$y, xx = object$x[, object$var.index])
      colnames(data.d)[1] <- "Response"
      colnames(data.d)[-1] <- object$var.label

      if (object$family == "gaussian") {
        summary(lm(Response ~ ., data = data.d))
      } else if (object$family == "binomial") {
        summary(glm(Response ~ .,
          family = "binomial", data = data.d,
          method = "brglmFit", type = "MPL_Jeffreys"
        ))
      } else {
        summary(glm(Response ~ ., family = "poisson", data = data.d))
      }
    } else {
      data.d <- data.frame(object$x[, object$var.index])
      colnames(data.d) <- object$var.label
      summary(coxph(Surv(object$y[, 1], object$y[, 2]) ~ ., data = data.d))
    }
  } else {
    message("None of variables are selected.")
    message("Therefore, the summary is shown for the model with intercept only\n")

    if (object$family != "cox") {
      data.d <- data.frame(yy = object$y)
      colnames(data.d)[1] <- "Response"
    }

    if (object$family == "gaussian") {
      summary(lm(Response ~ 1, data = data.d))
    } else if (object$family == "binomial") {
      summary(glm(Response ~ 1,
        family = "binomial", data = data.d,
        method = "brglmFit", type = "MPL_Jeffreys"
      ))
    } else if (object$family == "poisson") {
      summary(glm(Response ~ 1, family = "poisson", data = data.d))
    } else {
      summary(coxph(Surv(object$y[, 1], object$y[, 2]) ~ 1))
    }
  }
}

#' \code{predict.sgpv}: Prediction using the fitted model
#'
#' S3 method \code{predict} for an object of class \code{sgpv}
#'
#' @importFrom stats lm predict
#' @importFrom brglm2 brglmFit
#'
#' @param object An \code{sgpv} objectect
#' @param newdata Prediction data set
#' @param type The type of prediction required. Can take the value of `link`,
#' `response`, and `terms`. Default is `response`.
#' @param ... Other \code{predict} arguments
#'
#' @return Predicted values
#' @export
#' @examples
#'
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # run one-stage algorithm
#' out.sgpv <- pro.sgpv(x = x, y = y)
#'
#' predict(out.sgpv)
predict.sgpv <- function(object, newdata, type, ...) {
  if (object$family == "cox") stop("Sorry, prediction method isn't available for Cox models yet.\n")

  if (length(object$var.index) > 0) {
    data.d <- data.frame(Response = object$y, xx = object$x[, object$var.index])
    colnames(data.d)[-1] <- object$var.label

    if (missing(newdata)) newdata <- object$x
    if (is.null(colnames(newdata))) colnames(newdata) <- paste("V", 1:ncol(newdata), sep = "")
    if (missing(type)) type <- "response"

    if (object$family == "gaussian") {
      lm.m <- lm(Response ~ ., data = data.d)
      predict(lm.m, data.frame(newdata), ...)
    } else if (object$family == "poisson") {
      glm.m <- glm(Response ~ ., family = "poisson", data = data.d)
      predict(glm.m, data.frame(newdata), type = type, ...)
    } else {
      glm.m <- glm(Response ~ .,
        family = "binomial", data = data.d,
        method = "brglmFit", type = "MPL_Jeffreys"
      )
      predict(glm.m, data.frame(newdata), type = type, ...)
    }
  } else {
    message("None of variables are selected.")
    message("Therefore, the prediction is based on the intercept only model.\n")

    if (missing(newdata)) newdata <- object$x

    if (object$family == "gaussian") {
      data.d <- data.frame(yy = object$y)
      lm.m <- lm(yy ~ 1, data = data.d)
      predict(lm.m, data.frame(newdata), ...)
    } else {
      glm.m <- glm(yy ~ 1,
        data = data.frame(
          yy = object$y,
          xx = object$x
        ),
        family = object$family
      )

      predict(glm.m, newdata = data.frame(newdata), type = "response", ...)
    }
  }
}

#' \code{plot.sgpv}: Plot variable selection results
#'
#' S3 method \code{plot} for an object of class \code{sgpv}. When the two-stage
#' algorithm is used, this function plots the fully relaxed lasso solution path on
#' the standardized scale and the final variable selection results. When the
#' one-stage algorithm is used, a histogram of all coefficients with selected effects
#' is shown.
#'
#' @importFrom graphics abline axis lines mtext polygon
#' @importFrom graphics legend points
#' @param x An \code{sgpv} object
#' @param lpv Lines per variable. It can take the value of 1 meaning that only the
#' bound that is closest to the null will be plotted, or the value of 3 meaning that
#' point estimates as well as 95% confidence interval will be plotted. Default is 3.
#' @param lambda.max The maximum lambda on the plot. Default is \code{NULL}.
#' @param short.label An indicator if a short label is used for each variable for
#' better visualization. Default is \code{TRUE}
#' @param ... Other \code{plot} arguments
#'
#' @return NULL
#' @export
#' @examples
#'
#' # prepare the data
#' x <- t.housing[, -ncol(t.housing)]
#' y <- t.housing$V9
#'
#' # one-stage algorithm
#' out.sgpv.1 <- pro.sgpv(x = x, y = y, stage = 1)
#'
#' # plot the selection result
#'
#' plot(out.sgpv.1)
#'
#' # two-stage algorithm
#' out.sgpv.2 <- pro.sgpv(x = x, y = y)
#'
#' # plot the fully relaxed lasso solution path and final solution
#' plot(out.sgpv.2)
#'
#' # zoom in a little bit
#' plot(out.sgpv.2, lambda.max = 0.01)
#'
#' # only plot one confidence bound
#' plot(out.sgpv.2, lpv = 1, lambda.max = 0.01)
plot.sgpv <- function(x, lpv = 3, lambda.max = NULL, short.label = T, ...) {
  if (!lpv %in% c(1, 3)) stop("lpv argument only takes values of 1 and 3.")

  if (length(x$var.index) == 0) {
    message("None of variables are selected.\n")
    if (x$stage == 2) stop("No visualization is available.\n")
  }

  # get information from data
  p <- ncol(x$x)
  if (short.label == T) {
    x.names <- paste("V", 1:p, sep = "")
  } else {
    x.names <- colnames(x$x)
  }


  # standardize inputs in linear regression
  if (x$family == "gaussian") {
    xs <- scale(x$x)
    ys <- scale(x$y)
  } else {
    xs <- as.matrix(x$x)
    ys <- x$y
  }

  if (x$stage == 2) {

    # maximum lambda in gaussian
    if (is.null(lambda.max) & x$family == "gaussian") {
      lambda.max <- max(abs(sapply(1:ncol(xs), function(z) xs[, z] %*% ys)) / nrow(xs)) * 1.1
    } else if (is.null(lambda.max)) {
      lambda.max <- 0.5
    }


    step <- lambda.max / 100

    # get a sequence of lambda
    lambda.seq <- seq(0, lambda.max, step)

    # fit lasso once
    if (x$family != "cox") {
      lasso <- glmnet(xs, ys, family = x$family)
    } else {
      lasso <- glmnet(xs, Surv(ys[, 1], ys[, 2]), family = x$family)
    }

    # get coefficient estimates at each lambda
    if (p < length(x$y)) {
      results <- sapply(lambda.seq, function(z) {
        get.coef(
          xs = xs, ys = ys,
          lambda = z,
          lasso = lasso,
          family = x$family
        )
      })

      # prepare data to plot
      to.plot <- data.frame(
        lambda = rep(lambda.seq, each = p),
        v = rep(x.names, length(lambda.seq)),
        pe = c(results[1:p, ]),
        lb = c(results[(p + 1):(2 * p), ]),
        ub = c(results[(2 * p + 1):(3 * p), ])
      )
    } else {
      results <- sapply(lambda.seq[-1], function(z) {
        get.coef(
          xs = xs, ys = ys,
          lambda = z,
          lasso = lasso,
          family = x$family
        )
      })

      # prepare data to plot
      to.plot <- data.frame(
        lambda = rep(lambda.seq[-1], each = p),
        v = rep(x.names, length(lambda.seq) - 1),
        pe = c(results[1:p, ]),
        lb = c(results[(p + 1):(2 * p), ]),
        ub = c(results[(2 * p + 1):(3 * p), ])
      )
    }


    if (lpv == 1) {
      to.plot$bound <- ifelse(abs(to.plot$lb) < abs(to.plot$ub), to.plot$lb, to.plot$ub)
      plot.d <- to.plot[, c("lambda", "v", "bound")]

      # find the location of x.names
      location.beta <- to.plot$bound[to.plot$lambda == to.plot$lambda[1]]
    } else if (lpv == 3) {
      plot.d <- to.plot

      # find the location of x.names
      location.beta <- to.plot$pe[to.plot$lambda == to.plot$lambda[1]]
    }

    # get the indices of selected variables
    selected.index <- x$var.index
    black.index <- setdiff(1:p, selected.index)

    # change the color of the variables
    color.use <- c(
      rep("black", length(black.index)),
      rep("blue", length(selected.index))
    )

    location.beta <- location.beta[c(black.index, selected.index)]
    var.axis <- x.names[c(black.index, selected.index)]

    # find the limit of the canvas
    if (lpv == 3) {
      ylim <- c(
        min(c(location.beta, to.plot$lb, to.plot$ub)) * 1.01,
        max(c(location.beta, to.plot$lb, to.plot$ub)) * 1.01
      )

      ylim <- c(-max(abs(ylim)), max(abs(ylim)))
      ytick <- c(
        -round(max(abs(ylim)), 2), -round(max(abs(ylim)), 2) / 2, 0,
        round(max(abs(ylim)), 2) / 2, round(max(abs(ylim)), 2)
      )
    } else if (lpv == 1) {
      ylim <- c(
        min(plot.d$bound) * 1.01,
        max(plot.d$bound) * 1.01
      )

      ylim <- c(-max(abs(ylim)), max(abs(ylim)))
      ytick <- c(
        -round(max(abs(ylim)), 2), -round(max(abs(ylim)), 2) / 2, 0,
        round(max(abs(ylim)), 2) / 2, round(max(abs(ylim)), 2)
      )
    }

    # create a data set for the null bound
    n.bound <- results[(3 * p + 1), ]

    # find the lambda.gic
    vlambda <- x$lambda

    # plot the results
    if (lpv == 1) {
      xvals <- split(plot.d$lambda, plot.d$v)
      yvals <- split(plot.d$bound, plot.d$v)

      plot(lambda.seq,
        type = "n", ylim = ylim, xlim = c(0, lambda.max),
        xlab = expression(lambda), ylab = "Bound that is closer to the null",
        main = "Solution to the two-stage algorithm with one line per variable",
        axes = F, frame.plot = T, ...
      )
      axis(1, at = round(seq(0, lambda.max, length.out = 5), 3))
      axis(2, at = ytick, labels = c(ytick[1:2], 0, ytick[4:5]))
      mtext(var.axis, side = 2, at = location.beta, col = color.use)
      mtext(bquote(lambda["gic"]), side = 1, at = vlambda)
      # null region
      if (p < length(x$y)) {
        polygon(c(lambda.seq, rev(lambda.seq)), c(-n.bound, rev(n.bound)), col = "grey", border = "grey")
      } else {
        polygon(c(lambda.seq[-1], rev(lambda.seq[-1])), c(-n.bound, rev(n.bound)), col = "grey", border = "grey")
      }
      invisible(mapply(lines, xvals, yvals, col = 1:p))
      abline(h = 0)
      abline(v = vlambda, lty = 2)
    } else if (lpv == 3) {
      plot(lambda.seq,
        type = "n", ylim = ylim, xlim = c(0, lambda.max),
        xlab = expression(lambda), ylab = "Point estimates and confidence intervals",
        main = "Solution to the two-stage algorithm with three lines per variable",
        axes = F, frame.plot = T, ...
      )
      axis(1, at = round(seq(0, lambda.max, length.out = 5), 3))
      axis(2, at = ytick, labels = c(ytick[1:2], 0, ytick[4:5]))

      # null region
      if (p < length(x$y)) {
        polygon(c(lambda.seq, rev(lambda.seq)), c(-n.bound, rev(n.bound)), col = "grey", border = "grey")
      } else {
        polygon(c(lambda.seq[-1], rev(lambda.seq[-1])), c(-n.bound, rev(n.bound)), col = "grey", border = "grey")
      }

      # text on axis
      mtext(var.axis, side = 2, at = location.beta, col = color.use)
      mtext(bquote(lambda["gic"]), side = 1, at = vlambda)

      # point estimates
      xvals <- split(plot.d$lambda, plot.d$v)
      yvals <- split(plot.d$pe, plot.d$v)
      invisible(mapply(lines, xvals, yvals, col = 1:p))

      # upper bound
      xvals <- split(plot.d$lambda, plot.d$v)
      yvals <- split(plot.d$ub, plot.d$v)
      invisible(mapply(lines, xvals, yvals, col = 1:p, lty = 2))

      # lower bound
      xvals <- split(plot.d$lambda, plot.d$v)
      yvals <- split(plot.d$lb, plot.d$v)
      invisible(mapply(lines, xvals, yvals, col = 1:p, lty = 2))

      abline(h = 0)
      abline(v = vlambda, lty = 2)
    }
  } else if (length(x$var.index) > 0) { # one-stage algorithm

    var.names <- colnames(x$x)
    pe <- x$pe.can
    lb <- x$lb.can
    ub <- x$ub.can
    p <- ncol(x$x)
    null.bound <- x$null.bound

    col.names <- rep("black", p)
    col.names[x$var.index] <- "blue"

    # set the x-axis
    x.lower <- min(lb) * ifelse(min(lb) < 0, 1.1, 0.9)
    x.upper <- max(ub) * ifelse(max(ub) > 0, 1.1, 0.9)

    plot(0,
      type = "n", xlab = "Effect estimates", ylab = "Variables", yaxt = "n",
      xlim = c(x.lower, x.upper),
      ylim = c(0, p + 1),
      main = "Selection results in the one-stage algorithm", ...
    )
    mtext(paste("ProSGPV selects", length(x$var.index), "variables."), side = 3)
    lines(x = rep(null.bound, 2), y = c(0, p + 1), col = "green4")
    lines(x = -rep(null.bound, 2), y = c(0, p + 1), col = "green4")
    for (i in 1:p) {
      lines(x = c(lb[i], ub[i]), y = c(i, i), col = col.names[i])
      mtext(var.names[i], side = 2, at = i, col = col.names[i])
      points(x = pe[i], y = i, pch = 20, col = col.names[i])
    }
    legend("topleft",
      box.col = "white", box.lwd = 0,
      legend = c("ProSGPV selection", "Null bound"), col = c("blue", "green4"),
      lty = c(1, 1), cex = 0.8
    )
  } else {
    message("All effects in the full model are no different than or overlap with the null.\n")
  }
}
